### PR TITLE
Update In-Place Pod Resize docs for v1.32

### PR DESCRIPTION
**Pull request by tallclair**
### Description

Update the in-place pod resize documentation for ~beta in~ v1.32.

### Issue

For https://github.com/kubernetes/enhancements/issues/1287
Fixes https://github.com/kubernetes/website/issues/48534

### CODE REVIEW

**Comment by tallclair** [comment]
I will get this sent out today.

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 1]
```diff
@@ -229,12 +243,9 @@ spec:
...
 ```
 
-Observe that the `allocatedResources` values have been updated to reflect the new
-desired CPU requests. This indicates that node was able to accommodate the
-increased CPU resource needs.
-
-In the Container's status, updated CPU resource values shows that new CPU
-resources have been applied. The Container's `restartCount` remains unchanged,
+Observe that the `resources` in the `containerStatuses` have been updated to reflect the new desired
+CPU requests. This indicates that node was able to accommodate the increased CPU resource needs.c,
```
Comment (esotsal) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:247]: Unnecessary `.c`

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 2]
```diff
@@ -24,27 +24,32 @@ to be enabled. The alternative is to delete the Pod and let the
...
-  actual resource `requests` and `limits` that are configured on the running
-  containers as reported by the container runtime.
-- `resize` field in the Pod's status shows the status of the last requested
+- A container's resource `requests` and `limits` are _mutable_ for CPU
+  and memory resources. These fields represent the _desired_ resources for the container.
+- The `resources` field in `containerStatuses` of the Pod's status reflects the resources
+  _allocated_ to the pod's containers. For running containers, this reflects the actual resource
+  `requests` and `limits` that are configured as reported by the container runtime. For non-running
+  containers, these are the resources allocated for the container when it starts.
+- The `resize` field in the Pod's status shows the status of the last requested
```
Comment (esotsal) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:37]: Do you think, we should add a sentence when `resize` value is null ? For example if a pod was resized , and pending resize is null means that Kubelet has processed the resize ?

File: content/en/docs/concepts/workloads/autoscaling.md [hunk 3]
```diff
@@ -79,7 +79,7 @@ Mode | Description
 
 #### Requirements for in-place resizing
 
-{{< feature-state for_k8s_version="v1.27" state="alpha" >}}
+{{< feature-state feature_gate_name="InPlacePodVerticalScaling" >}}
```
Comment (tengqm) [inline content/en/docs/concepts/workloads/autoscaling.md:82]: We need to update the `/content/en/docs/reference/command-line-tools-reference/feature-gates/in-place-pod-vertical-scaling.md` file as well.

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md (hunk 1 above)
Comment (tallclair) [reply content/en/docs/tasks/configure-pod-container/resize-container-resources.md]: Thanks, fixed.

**Comment by tengqm** [comment]
Docs look good to me.
But I failed to find the upstream PR for this feature promotion.

File: content/en/docs/reference/command-line-tools-reference/feature-gates/in-place-pod-vertical-scaling.md [hunk 4]
```diff
@@ -9,5 +9,8 @@ stages:
   - stage: alpha
     defaultValue: false
     fromVersion: "1.27"
+  - stage: beta
+    defaultValue: true
+    fromVersion: "1.32"
 ---
```
Comment (esotsal) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/in-place-pod-vertical-scaling.md:15]: Is this a good place to add description of[ InPlacePodVerticalScalingExclusiveCPUs  and 
InPlacePodVerticalScalingAllocatedStatus](https://github.com/kubernetes/kubernetes/blob/c9092f69fc0c099062dd23cd6ee226bcd52ec790/pkg/features/kube_features.go#L284-L297) ?

**Comment by esotsal** [comment]
> Docs look good to me. But I failed to find the upstream PR for this feature promotion.

https://github.com/kubernetes/kubernetes/pull/128682

**Review by tengqm — COMMENTED** [review]
One nit. LGTM other than that.

**Comment by liggitt** [comment]
doc updates for behavior changes are fine, but drop out the beta promotion bit, beta promotion was reverted in https://github.com/kubernetes/kubernetes/pull/128875

**Comment by chanieljdan** [comment]
> doc updates for behavior changes are fine, but drop out the beta promotion bit, beta promotion was reverted in [kubernetes/kubernetes#128875](https://github.com/kubernetes/kubernetes/pull/128875)

Hi @tallclair as we approach docs freeze tomorrow, are you able to make this change? Docs freeze is tomorrow at 18:00 PDT.

**Comment by haircommander** [comment]
this KEP was pulled back to alpha state, @tallclair do you plan on updating 1.32 docs with what changed in alpha?

**Comment by tallclair** [comment]
The only thing that needs to change in this is reverting the feature state change. Everything else should be accurate. I won't be able to fix this until tomorrow though. If anyone wants to fork the pr that works be helpful, otherwise I'll need an exception to get to it tomorrow

File: content/en/docs/reference/command-line-tools-reference/feature-gates/in-place-pod-vertical-scaling.md [hunk 5]
```diff
@@ -9,5 +9,8 @@ stages:
   - stage: alpha
     defaultValue: false
     fromVersion: "1.27"
+  - stage: beta
+    defaultValue: true
+    fromVersion: "1.32"
```
Comment (sftim) [inline content/en/docs/reference/command-line-tools-reference/feature-gates/in-place-pod-vertical-scaling.md:14]:
```suggestion
```

**Comment by sftim** [comment]
How's [this](https://github.com/kubernetes/website/pull/48503#pullrequestreview-2462476426)?

**Comment by haircommander** [comment]
LGTM, I'm fine with this or https://github.com/kubernetes/website/pull/48857

**Comment by tallclair** [comment]
Changes lgtm. Thank you @sftim  for helping with this!

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 6]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
 [workload controller](/docs/concepts/workloads/controllers/) make a replacement Pod
 that has a different resource requirement.
 
+A resize request is made through the pod `/resize` subresource, which takes the full updated pod for
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:27]: nit: use Pod, not pod throughout this diff.

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 7]
```diff
@@ -107,6 +114,21 @@ have changed, the container will be restarted in order to resize its memory.
 
 <!-- steps -->
 
+## Limitations
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:117]: This belongs further up. Make this either a subheading of the prerequisites heading or a main heading before the prerequisites heading.

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 8]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
 [workload controller](/docs/concepts/workloads/controllers/) make a replacement Pod
 that has a different resource requirement.
 
+A resize request is made through the pod `/resize` subresource, which takes the full updated pod for
+an update request, or a patch on the pod object for a patch request.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:28]:
```suggestion
an update request, or a patch on the `Pod` object for a patch request.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 9]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
...
-- `resources` field in `containerStatuses` of the Pod's status reflects the
-  actual resource `requests` and `limits` that are configured on the running
-  containers as reported by the container runtime.
-- `resize` field in the Pod's status shows the status of the last requested
+- A container's resource `requests` and `limits` are _mutable_ for CPU
+  and memory resources. These fields represent the _desired_ resources for the container.
+- The `resources` field in `containerStatuses` of the Pod's status reflects the resources
+  _allocated_ to the pod's containers. For running containers, this reflects the actual resource
+  `requests` and `limits` that are configured as reported by the container runtime. For non-running
+  containers, these are the resources allocated for the container when it starts.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:36]:
```suggestion
- The `containerStatuses.resources` field in the Pod `status` field reflects the resources
   that are _allocated_ to the Pod's containers. For running containers, these are the actual values
   of `requests` and `limits`, as reported by the container runtime. For non-running containers,
   these are the resources that are allocated for when the container starts.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 10]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
...
+- The `resize` field in the Pod's status shows the status of the last requested
   pending resize. It can have the following values:
-  - `Proposed`: This value indicates an acknowledgement of the requested resize
-    and that the request was validated and recorded.
+  - `Proposed`: This value indicates that a pod was resized, but the Kubelet has not yet processed
+    the resize.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:40]:
```suggestion
  - `Proposed`: this value indicates that a Pod resize request was received, but the
     kubelet has not yet processed the resize.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 11]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
...
   - `InProgress`: This value indicates that the node has accepted the resize
     request and is in the process of applying it to the pod's containers.
   - `Deferred`: This value means that the requested resize cannot be granted at
     this time, and the node will keep retrying. The resize may be granted when
-    other pods leave and free up node resources.
+    other pods are removed and free up node resources.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:45]:
```suggestion
    other pods are removed and the available node resource capacity increases.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 12]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
...
   - `Infeasible`: is a signal that the node cannot accommodate the requested
     resize. This can happen if the requested resize exceeds the maximum
     resources the node can ever allocate for a pod.
+  - `""`: An empty or unset value indicates that the last resize completed. This should only be the
+    case if the resources in the container spec match the resources in the container status.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:50]:
```suggestion
  - `""`: An empty or unset value indicates that the most recent pending resize was completed.
     An empty or unset value only appears if the resources in the container specification match
     the resources in the container status.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 13]
```diff
@@ -24,26 +24,33 @@ to be enabled. The alternative is to delete the Pod and let the
...
+
+If a node has pods with an incomplete resize, the scheduler will compute the pod requests from the
+maximum of a container's desired resource requests, and it's actual requests reported in the status.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:53]:
```suggestion
If a node has Pods with an incomplete resize, the scheduler uses the largest of the following values to
calculate the Pod requests:

*  The desired resource requests for all containers in the Pod, as specified in the
   `spec.containers.resources.requests` field.
*  The actual resource requests for all containers in the Pod, as reported in the
   `spec.status.containerStatuses.resources.requests` field.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 14]
```diff
@@ -107,6 +114,21 @@ have changed, the container will be restarted in order to resize its memory.
 
 <!-- steps -->
 
+## Limitations
+
+In-place resize of pod resources currently has the following limitations:
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:119]:
```suggestion
In-place resize of Pod resources has the following limitations:
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 15]
```diff
@@ -107,6 +114,21 @@ have changed, the container will be restarted in order to resize its memory.
...
+
+- Only CPU and memory resources can be changed.
+- Pod QoS Class cannot change. This means that requests must continue to equal limits for Guaranteed
+  pods, Burstable pods cannot set requests and limits to be equal for both CPU & memory, and you
+  cannot add resource requirements to Best Effort pods.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:124]:
```suggestion
- The Pod QoS Class cannot change. The following considerations continue to apply:
   - Guaranteed QoS Pods: requests must be equal to limits.
   - Burstable QoS Pods: requests can't be equal to limits for both CPU and memory
   - Best Effort QoS Pods:  you can't set resource requirements
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 16]
```diff
@@ -107,6 +114,21 @@ have changed, the container will be restarted in order to resize its memory.
...
+- Init containers and Ephemeral Containers cannot be resized.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:125]:
```suggestion
- Init containers and ephemeral containers cannot be resized.
```

File: content/en/docs/tasks/configure-pod-container/resize-container-resources.md [hunk 17]
```diff
@@ -107,6 +114,21 @@ have changed, the container will be restarted in order to resize its memory.
...
+- Resource requests and limits cannot be removed once set.
+- A container's memory limit may not be reduced below its usage. If a request puts a container in
+  this state, the resize status will remain in `InProgress` until the desired memory limit becomes
+  feasible.
```
Comment (shannonxtreme) [inline content/en/docs/tasks/configure-pod-container/resize-container-resources.md:129]:
```suggestion
- A container's memory limit can't be reduced below the memory usage. If a request puts a 
  container in this state, the resize will remain in the `InProgress` state until the desired
  memory limit becomes  feasible.
```

**Comment by shannonxtreme** [comment]
@tallclair while I'd really like to see some of the editorial changes implemented before you submit, they aren't blocking, so I'm LGTMing as-is

/lgtm

**Comment by chanieljdan** [comment]
Docs and Technical LGTM noted. In the interest of time we'll approve these changes as is.

/approve

**Comment by reylejano** [comment]
The changes suggested by @shannonxtreme can be addressed in a follow-up PR
/approve

**Comment by mengjiao-liu** [comment]
It looks like the hold can now be lifted. @tallclair

**Comment by sftim** [comment]
/retitle Update In-Place Pod Resize docs for v1.32

**Comment by sftim** [comment]
Seems OK to unhold

If that's wrong, we can revert the change.
/hold cancel

**Comment by sftim** [comment]
nit cleanup PRs welcome!

**Pull request merged**